### PR TITLE
fix(core/webidl-index): support multiple idl blocks

### PR DESCRIPTION
**Pull request by shubhshrma**
For issue #1479 
The problem is appending only the first child.

Screenshot after changes: 

peek 2018-02-11 20-30

UPDATED SCREENSHOT

peek 2018-02-11 23-24

### CODE REVIEW

**Comment by marcoscaceres** [comment]
Great find! Could you please check why the idl-index test is now failing tho? We might’ve also want to add the test case from the original bug to the test suite.

File: src/core/webidl-index.js [hunk 1]
```diff
@@ -42,8 +42,11 @@ export function run(conf, doc, cb) {
   Array.from(document.querySelectorAll("pre.def.idl"))
     .map(elem => {
       const span = document.createElement("span");
-      const clone = elem.cloneNode(true).firstElementChild;
-      span.appendChild(clone);
+      const clone = elem.cloneNode(true).children;
```
Comment (marcoscaceres) [inline src/core/webidl-index.js]: Nit: `clone` is no longer s good name. Let’s go with:

```js
const [children] = element.cloneNode(true);
```

File: src/core/webidl-index.js [hunk 2]
```diff
@@ -42,8 +42,11 @@ export function run(conf, doc, cb) {
...
+      Array.from(clone).forEach(child => {
+        span.appendChild(child);
+        span.appendChild(document.createTextNode("\n"));
```
Comment (marcoscaceres) [inline src/core/webidl-index.js]: If a child is already a new line, should we won’t this give us multiple new lines?

Comment (shubhshrma) [reply src/core/webidl-index.js]: Yes. This is giving an extra newline after every ```pre.idl```. Looking into this.

File: src/core/webidl-index.js (hunk 1 above)
Comment (shubhshrma) [reply src/core/webidl-index.js]: I've updated the code. Please see whether it's good now.

**Comment by shubhshrma** [comment]
Regarding the tests, Travis passed now. I'll update the tests in some time @marcoscaceres.

**Comment by marcoscaceres** [comment]
Awesome! Thanks @shubhshrma.

**Comment by marcoscaceres** [comment]
Heh, autocorrect got me. I meant “quick”, not ”kick”, comments! :)

File: src/core/webidl-index.js [hunk 3]
```diff
@@ -42,9 +42,11 @@ export function run(conf, doc, cb) {
   Array.from(document.querySelectorAll("pre.def.idl"))
     .map(elem => {
       const span = document.createElement("span");
-      const clone = elem.cloneNode(true).firstElementChild;
-      span.appendChild(clone);
-      span.appendChild(document.createTextNode("\n"));
+      const children = elem.cloneNode(true).children;
+      Array.from(children).forEach(child => {
```
Comment (saschanaz) [inline src/core/webidl-index.js]: Can we use the new for-of syntax? `for (const child of Array.from(children))`

Comment (marcoscaceres) [reply src/core/webidl-index.js]: We’ve been using mostly functional style, but here it might make sense as it’s a single `forEach`.

**Comment by marcoscaceres** [comment]
@shubhshrma, when possible, please rebase.

**Comment by shubhshrma** [comment]
Sure I'll rebase @marcoscaceres. Um, I am facing a small problem writing tests in ```tests/spec/core/idl-index-spec.js```. I've written this:
```
it("allows multi-block idl", async () => {
    const body = `
      ${makeDefaultBody()}
      <section>
        <pre class=idl>
          [Constructor, Exposed=Window]
          interface BeforeInstallPromptEvent : Event {
              Promise&lt;PromptResponseObject&gt; prompt();
          };
          dictionary PromptResponseObject {
            AppBannerPromptOutcome userChoice;
          };
        </pre>
      </section>
      <section id="idl-index"></section>
    `;
    const expectedIDL = `[Constructor, Exposed=Window]
      interface BeforeInstallPromptEvent : Event {
          Promise&lt;PromptResponseObject&gt; prompt();
      };
      dictionary PromptResponseObject {
        AppBannerPromptOutcome userChoice;
      };\n`;
    const ops = {
      config: makeBasicConfig(),
      body,
    };
    const doc = await makeRSDoc(ops);
    var idlIndex = doc.querySelector("#idl-index");
    expect(idlIndex).not.toBe(null);
    expect(idlIndex.querySelector("pre").textContent).toEqual(expectedIDL);
  });
```
This is not passing because the entities ```&lt; &gt;``` are not escaped in ```expectedIDL```. Is there an escaping utility present?

**Comment by marcoscaceres** [comment]
> This is not passing because the entities &lt; &gt; are not escaped in expectedIDL. Is there an escaping utility present?

Just change them to "<" and ">", and that should work, no?

**Comment by marcoscaceres** [comment]
Sorry,  I mean change them in "expectedIDL".

**Comment by shubhshrma** [comment]
Please review @marcoscaceres @saschanaz. Sorry for the delay. I didn't realize that I had to make ```expectedIDL``` exactly formatted like ```body``` in the test. Now I have rebased and added the test.

File: tests/spec/core/idl-index-spec.js [hunk 4]
```diff
@@ -39,6 +39,42 @@ interface Bar {
...
+};\n`;
+    const ops = {
+      config: makeBasicConfig(),
+      body,
+    };
+    const doc = await makeRSDoc(ops);
+    var idlIndex = doc.querySelector("#idl-index");
+    expect(idlIndex).not.toBe(null);
+    console.log(idlIndex.querySelector("pre").textContent);
+    console.log(expectedIDL);
```
Comment (marcoscaceres) [inline tests/spec/core/idl-index-spec.js]: Please remove console statements

Comment (marcoscaceres) [reply tests/spec/core/idl-index-spec.js]: Better, convert these to expect statements.

File: tests/spec/core/idl-index-spec.js [hunk 5]
```diff
@@ -39,6 +39,42 @@ interface Bar {
...
+        [Constructor, Exposed=Window]
+        interface BeforeInstallPromptEvent : Event {
+            Promise&lt;PromptResponseObject&gt; prompt();
+        };
+        dictionary PromptResponseObject {
+          AppBannerPromptOutcome userChoice;
+        };
+        </pre>
+      </section>
+         <section id="idl-index"></section>
```
Comment (marcoscaceres) [inline tests/spec/core/idl-index-spec.js]: Nit: make sure tgis section aligns with the one above.

**Comment by shubhshrma** [comment]
Rebased and updated @marcoscaceres.

**Comment by marcoscaceres** [comment]
Just waiting on Travis to give ok.

**Comment by marcoscaceres** [comment]
@shubhshrma, please take a look at: https://github.com/w3c/respec/issues/1502

**Comment by shubhshrma** [comment]
Looking into it @marcoscaceres

**Comment by marcoscaceres** [comment]
going live now as v19.2.2

**Pull request merged**